book/example_7_echo_server.md

```diff
@@ -0,0 +1,211 @@
+#[tokio::main]
+async fn main() {
+    // parse command line arguments
+    let opts = Opts::parse();
+```
+
+After parsing the command line arguments, we get into invoking the client or server code. Before we do so, we set up some Rust-based networking. Specifically, in both cases we will need to allocate a UDP socket that is used for both sending and receiving messages. We do this by calling the async `bind_udp_socket` function, which is defined in `hydroflow/src/net.rs`. It is an async `future`, so requires appending `.await`; the function returns a pair of type `(UdpSink, UdpSource)`. These are the types that we'll use in Hydroflow to send and receive messages. (Note: your IDE might expand out the `UdpSink` and `UdpSource` traits to their more verbose definitions. This is fine; you can ignore for now.)
+
+For the server case, all that's left is to invoke `run_server` and pass it the network information. Note that the server is also an asynchronous program, so we append `.await` to that call as well. The program will block on this call until the server is done (which should only happen when it fails).
+```rust,ignore
+    // depending on the role, pass in arguments to the right function
+    match opts.role {
+        Role::Server => {
+            // allocate `outbound` and `inbound` sockets
+            let (outbound, inbound) = bind_udp_socket(opts.server_addr.clone()).await;
+            // run the server
+            run_server(outbound, inbound, opts.graph.clone()).await;
+        }
+```
+
+In the client case, we need one more piece of information passed down: the address of the server. We get this by calling the `ipv4_resolve` function, which is also defined in `src/net.rs`. This function takes a string and returns a `SocketAddr` type, which is the type that the `UdpSink` and `UdpSource` traits expect. Invoking `run_client` is similar to the server case, except that we pass in the server address as well.
+```rust,ignore
+        Role::Client => {
+            // resolve the server's IP address
+            let server_addr = ipv4_resolve(opts.server_addr.clone());
+            // allocate `outbound` and `inbound` sockets
+            let (outbound, inbound) = bind_udp_socket(opts.addr.clone().unwrap()).await;
+            // run the client
+            run_client(outbound, inbound, server_addr).await;
+        }
+    }
+}
+```
+
+## protocol.rs
+As a design pattern, it is natural in distributed Hydroflow programs to define various message types in a `protocol.rs` file with structures shared for use by all the Hydroflow logic across roles. In this simple example, we define only one message type: `EchoMsg`, and a simple struct with two fields: `payload` and `ts` (timestamp). The `payload` field is a string, and the `ts` field is a `DateTime<Utc>`, which is a type from the [`chrono`](https://docs.rs/chrono/latest/chrono/) crate. Note the various derived traits on `EchoMsg` -- these are required for structs that we send over the network.
```

Comment:
Note the various derived traits on `EchoMsg`, specifically `Serialize` and `Deseriailize` --